### PR TITLE
feat(storage): use timeout for filter_extractor acquire

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

as title 

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- use timeout for notify to replace the unbound wait for debug

## Checklist

- ~[ ] I have written necessary rustdoc comments~
- ~[ ] I have added necessary unit tests and integration tests~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- https://github.com/risingwavelabs/risingwave/issues/5437

### CODE REVIEW

**Comment by Li0k** [comment]
when timeout

```
thread 'risingwave-main' panicked at 'called `Result::unwrap()` on an `Err` value: Elapsed(())', /Users/li0k/Documents/github/risingwave/src/storage/hummock_sdk/src/filter_key_extractor.rs:304:66
```

**Comment by fuyufjh** [comment]
Can you provide some context? 👀

**Comment by Li0k** [comment]
> Can you provide some context? 👀

When creating mv, meta will push the catalog to compute and compactor. When the system has acuqire timeout, it will be a bug (which will cause unbounded wait in main branch), we can find the problem in time by panic over timeout. I simulate the scenario above by breaking the notify code.

File: src/storage/hummock_sdk/src/filter_key_extractor.rs [hunk 1]
```diff
@@ -23,9 +24,12 @@ use risingwave_common::util::ordered::OrderedRowDeserializer;
 use risingwave_common::util::sort_util::OrderType;
 use risingwave_pb::catalog::Table;
 use tokio::sync::Notify;
+use tokio::time::timeout;
 
 use crate::key::{get_table_id, TABLE_PREFIX_LEN};
 
+pub static ACQUIRE_TIMEOUT: LazyLock<Duration> = LazyLock::new(|| Duration::from_secs(60));
```
Comment (BugenZhao) [inline src/storage/hummock_sdk/src/filter_key_extractor.rs:31]: `Duration::from_secs` is a `const fn`, so I believe it's okay to make it a constant directly.

```suggestion
const ACQUIRE_TIMEOUT: Duration = Duration::from_secs(60);
```

File: src/storage/hummock_sdk/src/filter_key_extractor.rs [hunk 2]
```diff
@@ -298,7 +302,7 @@ impl FilterKeyExtractorManagerInner {
             }
 
             if !table_id_set.is_empty() {
-                notified.await;
+                timeout(*ACQUIRE_TIMEOUT, notified).await.unwrap();
```
Comment (BugenZhao) [inline src/storage/hummock_sdk/src/filter_key_extractor.rs:305]: How about adding more descriptions with `.expect(..)`?

File: src/storage/hummock_sdk/src/filter_key_extractor.rs (hunk 1 above)
Comment (Li0k) [reply src/storage/hummock_sdk/src/filter_key_extractor.rs]: Thanks fix it

File: src/storage/hummock_sdk/src/filter_key_extractor.rs [hunk 3]
```diff
@@ -298,7 +302,14 @@ impl FilterKeyExtractorManagerInner {
             }
 
             if !table_id_set.is_empty() {
-                notified.await;
+                timeout(ACQUIRE_TIMEOUT, notified)
```
Comment (BugenZhao) [inline src/storage/hummock_sdk/src/filter_key_extractor.rs:305]: How about only enabling this on `#[cfg(debug_assertions)]` as we panic here. I'm not sure what will happen in production. 😢

**Review by BugenZhao — COMMENTED** [review]
LGTM

File: src/storage/hummock_sdk/src/filter_key_extractor.rs (hunk 3 above)
Comment (Li0k) [reply src/storage/hummock_sdk/src/filter_key_extractor.rs:305]: > How about only enabling this on `#[cfg(debug_assertions)]` as we panic here? I'm not sure what will happen in production. 😢 cc @yezizp2012

If stuck bug happen in production without panic, its difficult for us to check it ? 

😢 how about introduce the `async-stack-trace` in production and panic on `debug_assertions ` ?

Comment (BugenZhao) [reply src/storage/hummock_sdk/src/filter_key_extractor.rs:305]: Yes, we've planned to enable the async stack tracing in production.

Comment (yezizp2012) [reply src/storage/hummock_sdk/src/filter_key_extractor.rs:305]: If stuck, a recovery progress should have been triggered already if it's not caused by any bugs. I think panic here is acceptable (only on `#[cfg(debug_assertions)]`), and I think eventually we should have a way to notice this happened in production. An alarm system based on logs may helps.

**Review by yezizp2012 — APPROVED** [review]
LGTM, let make it only panic on `#[cfg(debug_assertions)]` firstly.

**Comment by Li0k** [comment]
@BugenZhao PTAL

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**